### PR TITLE
Add Linux Consumption Testing Utils

### PR DESCRIPTION
**Pull request by Hazhzeng**
## Description

This new `testutils_lc.py` allows end-to-end testing scenario to run through the Linux Consumption Container => Placeholder Mode => Specialization => Function Load => Function Invocation.

It ensures any gaps between linux consumption specialization is well covered.

The example apps are not lying in the repository, it is in the following blob storage (read-only):
https://pythonworker36sa.blob.core.windows.net/python-worker-lc-apps?restype=container&comp=list
https://pythonworker37sa.blob.core.windows.net/python-worker-lc-apps?restype=container&comp=list
https://pythonworker38sa.blob.core.windows.net/python-worker-lc-apps?restype=container&comp=list
https://pythonworker39sa.blob.core.windows.net/python-worker-lc-apps?restype=container&comp=list

When failure occurs, it also prints out the error messages from the containers for fast diagnoses.

To add more scenarios:
1. Create a function app locally
2. Use `func azure functionapp publish python-worker-<python_version>-lc-func` to remote build the artifact (e.g. python-worker-37-lc-func)
3. Download the artifact from `https://python-worker-<python_version>-lc-func.scm.azurewebsites.net/api/settings` => `SCM_RUN_FROM_PACKAGE`.
4. Upload the package onto `Simple Batch` Subscription => `AzureFunctionsPythonWorkerCILinuxDevOps` Resource Group => `pythonworker<python_major><python_minor>sa` Storage Account => `python-worker-lc-apps` Blob Container
5. Follow the example in `endtoend/test_linux_consumption.py` to create a new test case.
/cc: @anatoliB @vrdmr @stefanushinardi 

---

### PR information

- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [x] New Unit tests were added for the changes made and CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-functions-python-worker/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -396,6 +396,7 @@ def run(self):
             'azure-functions==1.7.2',
             'azure-eventhub~=5.1.0',
             'python-dateutil~=2.8.1',
+            'pycryptodome~=3.10.1',
```
Comment (Hazhzeng) [inline setup.py:399]: This library is used for generating **encryptedContext** when specializing a placeholder in Linux Consumption container. See `_encrypt_context` implementation.

**Comment by Hazhzeng** [comment]
GitHub Action host agent is not able to find `docker` or `/usr/bin/docker` in the python (pytest) process.
```
FileNotFoundError: [Errno 2] No such file or directory: '/usr/bin/docker run -p 0:80 -d --name adc7230d-fc37-4278-8026-13a479deb455 --cap-add SYS_ADMIN --device /dev/fuse -e "CONTAINER_NAME"="adc7230d-fc37-4278-8026-13a479deb455" -e "CONTAINER_ENCRYPTION_KEY"="MDEyMzQ1Njc4OUFCQ0RFRjAxMjM0NTY3ODlBQkNERUY=" -e "WEBSITE_PLACEHOLDER_MODE"="1" -v "/home/runner/work/azure-functions-python-worker/azure-functions-python-worker/azure_functions_worker":"/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker" mcr.microsoft.com/azure-functions/mesh:3.1.4-upgrade': '/usr/bin/docker run -p 0:80 -d --name adc7230d-fc37-4278-8026-13a479deb455 --cap-add SYS_ADMIN --device /dev/fuse -e "CONTAINER_NAME"="adc7230d-fc37-4278-8026-13a479deb455" -e "CONTAINER_ENCRYPTION_KEY"="MDEyMzQ1Njc4OUFCQ0RFRjAxMjM0NTY3ODlBQkNERUY=" -e "WEBSITE_PLACEHOLDER_MODE"="1" -v "/home/runner/work/azure-functions-python-worker/azure-functions-python-worker/azure_functions_worker":"/azure-functions-host/workers/python/3.6/LINUX/X64/azure_functions_worker" mcr.microsoft.com/azure-functions/mesh:3.1.4-upgrade'
```

**Comment by Hazhzeng** [comment]
Able to run docker with command, but failed to mount in mesh container.
MS_FUNCTION_LOGS 2,,mesh-placeholder,mesh-placeholder,mesh.init,Host.Startup,"","Request for / failed with err exit status 1 error: **fuse: mount failed: Permission denied**  output: ",3.1.4.0,2021-08-04T23:04:14.140267986Z,,"",,,,77969af3-87bc-4904-af12-838e2888bb78,,

**Comment by Hazhzeng** [comment]
Running the docker container with `--priviledged` flag resolves all the issues. This PR is ready for review now.

File: azure_functions_worker/testutils_lc.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+    @property
+    def url(self) -> str:
+        if self._uuid not in self._ports:
+            raise RuntimeError(f'Failed to assign container {self._name} since'
+                               ' it is not spawned')
+
+        return f'http://localhost:{self._ports[self._uuid]}'
+
+    def assign_container(self, env: Dict[str, str] = {}):
```
Comment (balag0) [inline azure_functions_worker/testutils_lc.py:52]: This is great. How long do we wait after the container starts before sending the /assign request?

Comment (Hazhzeng) [reply azure_functions_worker/testutils_lc.py:52]: It waits for 3 seconds.

File: setup.py (hunk 1 above)
Comment (Hazhzeng) [reply setup.py:399]: License BSD: https://github.com/Legrandin/pycryptodome/blob/master/LICENSE.rst#bsd-license

File: azure_functions_worker/testutils_lc.py (hunk 2 above)
Comment (balag0) [reply azure_functions_worker/testutils_lc.py:52]: great. that should be good enough.

File: azure_functions_worker/testutils_lc.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+from Crypto.Hash.SHA256 import SHA256Hash
+from Crypto.Util.Padding import pad
+import requests
+
+# Linux Consumption Testing Configs
+DOCKER_PATH = "DOCKER_PATH"
+DOCKER_DEFAULT_PATH = "docker"
+MESH_IMAGE_URL = "https://mcr.microsoft.com/v2/azure-functions/mesh/tags/list"
+MESH_IMAGE_REPO = "mcr.microsoft.com/azure-functions/mesh"
+DUMMY_CONT_KEY = "MDEyMzQ1Njc4OUFCQ0RFRjAxMjM0NTY3ODlBQkNERUY="
```
Comment (vrdmr) [inline azure_functions_worker/testutils_lc.py:25]: Things which are doing to be constant and not changing, add an '_' to then as prefix.

File: azure_functions_worker/testutils_lc.py (hunk 2 above)
Comment (vrdmr) [reply azure_functions_worker/testutils_lc.py:52]: Is there a way we can call an API in the container to check if it is active and responding? Such that we don't have to add an artificial wait?

File: azure_functions_worker/testutils_lc.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+        # Add compulsory fields in specialization context
+        env["FUNCTIONS_EXTENSION_VERSION"] = f"~{self._host_version}"
+        env["FUNCTIONS_WORKER_RUNTIME"] = "python"
+        env["FUNCTIONS_WORKER_RUNTIME_VERSION"] = self._py_version
+        env["WEBSITE_SITE_NAME"] = self._uuid
+        env["WEBSITE_HOSTNAME"] = f"{self._uuid}.azurewebsites.com"
+
+        # Send the specialization context via a POST request
+        response = requests.post(
```
Comment (vrdmr) [inline azure_functions_worker/testutils_lc.py:66]: I saw a `send_request` below - why aren't you using that one here?

**Review by vrdmr — COMMENTED** [review]
review in progress - will complete by tomorrow.

File: azure_functions_worker/testutils_lc.py (hunk 4 above)
Comment (Hazhzeng) [reply azure_functions_worker/testutils_lc.py:66]: Good catch! Corrected!

**Review by vrdmr — APPROVED** [review]

**Pull request merged**